### PR TITLE
Workaround for duplicate resource http_port #120

### PR DESCRIPTION
**Pull request by granquet**
This is a rebase ontop of master for: https://github.com/voxpupuli/puppet-squid/pull/135
The review comment from @ekohl has been adressed in the process.

The module is now able to handle multiple server
declarations for the same port on different IPs.

### CODE REVIEW

**Comment by granquet** [comment]
I've added a minor correction (unrelated to the topic of the PR though) in the way the check is made for selinux.
on the two systems I've tried:
- CentOS Linux release 8.1.1911 (Core) with facter 3.14.4
$facts['os']['selinux'] is a hash and thus testing for true doesn't seem right (and doesn't work)
- Scibian 9 ( Debian 9.13 ) with facter 2.4.6
the selinux key in $facts['os'] doesn't exists.and testing for $facts['selinux'] seems frowned upon

my best bet seems to be: has_key($facts['os'],'selinux') and $facts['os']['selinux']['enabled'] == true
It makes rake/travis happy, works on my old debian (without selinux) and on CentOS (with selinux).

I'm really new to ruby and puppet in general... so, not sure of what I'm doing :)
please advise?

File: manifests/cache_dir.pp [hunk 1]
```diff
@@ -54,7 +54,7 @@
     }
   }
 
-  if $facts['os']['selinux'] == true {
+  if has_key($facts['os'],'selinux') and $facts['os']['selinux']['enabled'] == true {
```
Comment (ekohl) [inline manifests/cache_dir.pp:57]: I think this is equally effective.
```suggestion
  if fact('os.selinux.enabled') {
```

However, This should also always work
```suggestion
  if $facts['os']['selinux']['enabled'] {
```

(comment is also valid in other files, but I won't repeat the same comment multiple times)

File: spec/classes/init_spec.rb [hunk 2]
```diff
@@ -513,15 +513,15 @@
...
+      if facts[:osfamily] == 'RedHat'
         context 'with http_port parameters set + SELINUX' do
           let :params do
             { config: '/tmp/squid.conf',
               http_ports: { 2000 => { 'options' => 'special for 2000' } } }
           end
           let(:facts) do
             facts.merge(
-              selinux => true
+              selinux: true
```
Comment (ekohl) [inline spec/classes/init_spec.rb:524]: This sets the legacy fact. See https://github.com/voxpupuli/voxpupuli-test#fact-handling for how we have `override_facts` to correctly set modern structured facts.

File: spec/classes/init_spec.rb [hunk 3]
```diff
@@ -513,15 +513,15 @@
         it { is_expected.to contain_concat_fragment('squid_https_port_2001').with_content(%r{^https_port\s+2001\s+special for 2001$}) }
       end
 
-      if facts['osfamily'] == 'RedHat'
+      if facts[:osfamily] == 'RedHat'
```
Comment (ekohl) [inline spec/classes/init_spec.rb:516]: Good catch.

File: manifests/cache_dir.pp (hunk 1 above)
Comment (ekohl) [reply manifests/cache_dir.pp:57]: I just read over that you want to support Facter 2.4. In that case [fact()](https://forge.puppet.com/puppetlabs/stdlib/reference#fact) is the best thing to use.

**Comment by bastelfreak** [comment]
@ekohl do we really want to support facter 2? This isn't really maintained by Puppet since a long time and we dropped that in many many modules. Also I think this makes the code harder to read. @granquet is there a reason for not using the Puppetlabs repositories for modern puppet/facter? We don't officially support Puppet 4 / facter 2 anymore.

**Comment by ekohl** [comment]
I think `if fact('os.selinux.enabled') {}` is not harder to read than `if $fact['os']['selinux']['enabled'] {}`. There may be cases where it's less clear, but in this case there is no fallback to support facter 2. IMHO it's fine to keep it that way if it makes a user happy.

**Comment by bastelfreak** [comment]
okay, fine by me.

**Comment by granquet** [comment]
I would say that we have installed the machines with puppet 4/facter 2 and we will stick with it until the end of life of the machines.
the next machine we are installing are running puppet 6/facter 3.

let me drop that PR, as @ralfbosz updated the original one I was intending to get merged with this one.
https://github.com/voxpupuli/puppet-squid/pull/135

**Pull request closed without merging**